topics/go/algorithms/data/hash/hash.go

```diff
@@ -0,0 +1,107 @@
+// Package hash implements a hash table
+package hash
+
+import (
+	"fmt"
+	"hash/maphash"
+)
+
+const (
+	numBuckets = 256 // Number of buckets in hash
+)
+
+// entry in the hash table
+type entry struct {
+	key   string
+	value int
+}
+
+// Hash is a simple Hash table
+type Hash struct {
+	buckets [][]*entry
+	h       maphash.Hash
+}
+
+// NewHash returns a new hash table
+func NewHash() *Hash {
+	return &Hash{
+		buckets: make([][]*entry, numBuckets),
+	}
+}
+
+// Set sets the key to value in the hash
+func (h *Hash) Set(key string, value int) {
+	b := h.hashKey(key)
+	// Look for existing key
+	for _, e := range h.buckets[b] {
+		if e.key == key {
+			e.value = value
+			return
+		}
+	}
+
+	h.buckets[b] = append(h.buckets[b], &entry{key, value})
+}
+
+// Get gets the value associated with key, return a error is key not found
+func (h *Hash) Get(key string) (int, error) {
+	b := h.hashKey(key)
+	for _, e := range h.buckets[b] {
+		if e.key == key {
+			return e.value, nil
+		}
+	}
+
+	return 0, fmt.Errorf("%q not found", key)
+}
+
+// Delete deletes an entry from the hash, return an error if not found
+func (h *Hash) Delete(key string) error {
+	b := h.hashKey(key)
+	bucket := h.buckets[b]
+	for i, e := range bucket {
+		if e.key == key {
+			copy(bucket[i:], bucket[i+1:])
+			bucket = bucket[:len(bucket)-1]
+			// Free memory when bucket shrinks a lot
```

Comment:
Line space, proper sentence structure.

Also, this is teaching material, we need more comments walking the reader through the code. Assume someone looking at code like this for the first time. I know it will make the functions longer, but the purpose is to teach.

Reply:
OK